blob/fileblob/fileblob.go

```diff
@@ -565,6 +582,20 @@ func (b *bucket) Delete(ctx context.Context, key string) error {
 }
 
 func (b *bucket) SignedURL(ctx context.Context, key string, opts *driver.SignedURLOptions) (string, error) {
-	// TODO(Issue #546): Implemented SignedURL for fileblob.
+	path, err := b.path(key)
+	if err != nil {
+		return "", err
+	}
+
+	return b.urlSigner.SignedURL(ctx, key, path, opts)
+}
+
+type URLSigner interface {
+	SignedURL(ctx context.Context, key string, path string, opts *driver.SignedURLOptions) (string, error)
```

Comment:
It's not clear to me what we're asking the provider of this interface to do here. It looks like "everything". Can't we share more of the implementation? For example, we're going to want to look at `opts.Expiry` and include that in the URL somewhere, and then when we're serving the URL we'll want to verify that the expiration time is in the future. There's no reason to ask the user to implement that.

So, what are the things that we can reasonably expect a user to want to control? At least:

- The base URL / URL prefix for URLs.
- The signing mechanism (e.g., given a URL, produce a signature).
- The hook for handling incoming HTTP requests.

We can provide a default for the signing mechanism but not the other two. So maybe something like:

```
type Options struct {
  SignedURLPrefix string  // required to use SignedURL
  SignedURLHandler func(...) // required to use SignedURL
  SignedURLSigner func(u *url.URL) ([]byte, error) // defaults to MD5 hash or something
}
```

Reply:
I think the approach here is [SOLID](https://en.wikipedia.org/wiki/SOLID) (pun intended) in that it allows different URL construction schemes (both lexical or perhaps involving RPCs) and makes the fileblob bucket only responsible for serving content. A bucket is then composed of an interchangeable signer that handles the details of signing, which could in fact have the knobs you suggest.